exotations/task_maps/task_map/src/JointAccBackFiniteDiff.cpp

```diff
@@ -0,0 +1,101 @@
+ * Copyright (c) 2018, University Of Edinburgh
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *  * Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *  * Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *  * Neither the name of  nor the names of its contributors may be used to
+ *    endorse or promote products derived from this software without specific
+ *    prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include "task_map/JointAccBackFiniteDiff.h"
+
+REGISTER_TASKMAP_TYPE("JointAccBackFiniteDiff", exotica::JointAccBackFiniteDiff);
+
+namespace exotica
+{
+// Assign scene and initialize
+void JointAccBackFiniteDiff::assignScene(Scene_ptr scene)
+{
+    scene_ = scene;
+    Initialize();
+}
+
+// Instantiate
+void JointAccBackFiniteDiff::Instantiate(JointAccBackFiniteDiffInitializer& init)
+{
+    init_ = init;
+}
+
+// Initialize task map.
+//  Sets N (number of dofs), I_ used to set the Jacobian in update(), and e.
+//  e is the constant vector (-2, 1) used to approximate the joint acceleration using backwards finite differencing.
+void JointAccBackFiniteDiff::Initialize()
+{
+    N = scene_->getSolver().getNumControlledJoints();
+    I_ = Eigen::MatrixXd::Identity(N, N);
+    e << -2, 1;
+}
+
+// Initialize Q_
+//   Q_ is a log of the previous three joint states.
+//   Q_.col(0) is the most recent joint state.
+//   initPrevJointState sets each column of Q_ to joint_state.
+void JointAccBackFiniteDiff::initPrevJointState(Eigen::VectorXdRefConst joint_state)
+{
+    if (joint_state.rows() != N) throw_named("Wrong size for joint_state!");
+    Q_.resize(N, 2);
+    Q_.col(0) = joint_state;
+    Q_.col(1) = joint_state;
+    Qe_ = Q_ * e;
+}
+
+// Sets the previous joint state.
+//   In users script (python/c++) this must be called preceeding every time solve.solve() is called.
```

Comment:
This comment highlights an implicit assumption: Since you do not check whether it has been initialised or not you can create nasty and hard to debug segfaults due to differing matrix sizes etc. which mess up the whole Phi/Jacobian structures. Please always have a default initialisation, check whether it is fully initialised etc. One solution can be to initialise to zero so it will still work but initially not smooth for first few optimisations. Requiring these hidden assumptions isn't the way to go though.